tox_pdm/plugin4.py

```diff
@@ -0,0 +1,232 @@
+        groups = self.conf["groups"]
+        cmd = ["pdm", "install", "--no-self"]
+        for group in groups:
+            cmd.extend(("--group", group))
+        self.execute(cmd, StdinSource.OFF)
+        bin_path = self.env_site_package_dir() / "bin"
+        if not bin_path.exists():
+            return
+        scripts_path = self.env_bin_dir()
+        for file in bin_path.iterdir():
+            shutil.move(os.fspath(file), os.fspath(scripts_path))
+
+    def register_config(self) -> None:
+        super().register_config()
+        self.conf.add_config(
+            "groups",
+            of_type=t.List[str],
+            default=[],
+            desc="Specify the dependency groups to install",
+        )
+
+    @property
+    def _default_package_tox_env_type(self) -> str:
+        return "pdm-pep-517"
+
+    @staticmethod
+    def id() -> str:
+        return "pdm"
+
+
+class PdmPackageEnv(Pdm, PackageToxEnv):
+    def register_config(self) -> None:
+        super().register_config()
+        self.conf.add_config(
+            keys=["pkg_dir"],
+            of_type=Path,
+            default=lambda conf, name: self.env_dir / "dist",
+            desc="directory where to put project packages",
+        )
+
+    @property
+    def pkg_dir(self) -> Path:
+        return t.cast(Path, self.conf["pkg_dir"])
+
+    def perform_packaging(self, for_env: EnvConfigSet) -> t.List[Package]:
+        of_type: str = for_env["package"]
+        cmd = [
+            "pdm",
+            "build",
+            "-p",
+            str(self.conf["package_root"]),
+            "-d",
+            str(self.pkg_dir),
+        ]
+        if of_type == "wheel":
+            cmd.append("--no-sdist")
+            suffix = ".whl"
+        else:
+            cmd.append("--no-wheel")
+            suffix = ".tar.gz"
+        self.execute(cmd, StdinSource.OFF)
+        path = next(self.pkg_dir.glob(f"*{suffix}"))
+        package = WheelPackage(path, [])
+        return [package]
+
+    @staticmethod
+    def id() -> str:
+        return "pdm-pep-517"
+
+
+class PipPep582(Pip):
+    def installed(self) -> t.List[str]:
+        cmd = ["pdm", "list", "--freeze"]
+        result = self._env.execute(
+            cmd=cmd,
+            stdin=StdinSource.OFF,
+            run_id="freeze",
+            show=self._env.options.verbosity > DEFAULT_VERBOSITY,
+        )
+        result.assert_success()
+        return [line.strip() for line in result.out.splitlines() if line.strip()]
+
+
+class PDMRunExecutor(Execute):
+    def build_instance(
+        self,
+        request: ExecuteRequest,
+        options: ExecuteOptions,
+        out: SyncWrite,
+        err: SyncWrite,
+    ) -> "ExecuteInstance":
+        return PDMRunExecuteInstance(request, options, out, err)
+
+
+class PDMRunExecuteInstance(LocalSubProcessExecuteInstance):
+    @property
+    def cmd(self) -> t.Sequence[str]:
+        if self._cmd is None:
+            pdm_exe = self.options._env.options.pdm
```

Comment:
You're accessing here a private field 🤔 This can break without warning.

Reply:
The solution here might be that `options` needs to be passed for the execution request (in core).